shimmingtoolbox/shim/sequencer.py

```diff
@@ -830,25 +831,37 @@ def select_optimizer(method, unshimmed, affine, coils: ListCoil, pmu: PmuResp =
     return optimizer
 
 
-def _optimize(optimizer: Optimizer, nii_mask_anat, slices_anat, shimwise_bounds=None,
+_optimize_scope = {}  # i?: (_optimizer: Optimizer, nii_mask_anat, slices_anat, shimwise_bounds=None,
+                      #     dilation_kernel='sphere', dilation_size=3, path_output=None) 
+def _optimize(_optimizer: Optimizer, nii_mask_anat, slices_anat, shimwise_bounds=None,
               dilation_kernel='sphere', dilation_size=3, path_output=None):
 
     # Count shims to perform
     n_shims = len(slices_anat)
+    
+    global _optimize_scope
+    global optimizer
+    optimizer = _optimizer # not a super safe thing to do
 
     # multiprocessing optimization
-    mp.set_start_method('spawn', force=True)
+    mp.set_start_method('fork', force=True)
+    run_id = uuid.uuid4().hex
+    _optimize_scope[run_id] = (optimizer, nii_mask_anat, slices_anat, dilation_kernel, dilation_size, path_output, shimwise_bounds)
```

Comment:
I wonder if you can use a closure instead of a global:

```suggestion
    def __opt(i):
        return  _opt(i, optimizer, nii_mask_anat, slices_anat, dilation_kernel, dilation_size, path_output, shimwise_bounds)
```

And then below call 

```
         results = pool.starmap_async(__opt, [i for i in range(n_shims)]).get()
```

((and we would want to rename the functions for clarity))

By doing this, I think you should get the same benefit: at the point of `os.fork()`, `optimizer` and `nii_mask_anat` and all the rest have been stored in the [`__opt`'s closure](https://docs.python.org/3/library/inspect.html#inspect.getclosurevars) and will get "zero-copied" into each subprocess, and you don't need `global` or `uuid`.

But I'm not sure and I haven't tested. `Pool.apply_async` might try to pickle the `func` you pass it. I tried to read through multiprocessing's code and got lost. I think we should just try it on another-other branch and see how it affects the runtime.

Reply:
I tried it and the function being called by `pool.starmap_async()` needs to be a "top level function". See [here](https://stackoverflow.com/questions/52265120/python-multiprocessing-pool-attributeerror). With that in mind, I don't think this solution works.

Reply:
I tried it and the function being called by `pool.starmap_async()` needs to be a "top level function". See [here](https://stackoverflow.com/questions/52265120/python-multiprocessing-pool-attributeerror). With that in mind, I don't think this solution works.

Reply:
Ah cool, thanks for actually trying it out! I was going to get around to trying it at some point.

I forgot about the top-level restriction; okay, in that case, I think this weird pseudo-global thing is the best option.

We could maybe attach the values to that top-level function itself:

```
_opt.nii_mask_anat = nii_mask_anat
```

and then so long as the subprocesses never recurse it should be okay, and we can skip `run_id`. But maybe that's more fragile than it's worth.